control/setup_control.go

```diff
@@ -0,0 +1,399 @@
+package control
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/logging"
+	rdkutils "go.viam.com/rdk/utils"
+)
+
+/*
+SetupControlLoop
+tunePID
+GetState ?
+SetState ?
+*/
+
+// rPiGain is 1/255 because the PWM signal on a pi (and most other boards)
+// is limited to 8 bits, or the range 0-255.
+const rPiGain = 0.00392157
+
+var (
+	// default derivative type is "backward1st1".
+	derivativeType   = "backward1st1"
+	loopFrequency    = 50.0
+	sumIndex         = 1
+	linearPIDIndex   = 2
+	angularPIDIndex  = -1
+	controllableType = "motor_name"
+)
+
+// PIDLoop is used for setting up a PID control loop.
+type PIDLoop struct {
+	BlockNames              map[string][]string
+	ControlConf             Config
+	ControlLoop             *Loop
+	Options                 Options
+	Controllable            Controllable
+	logger                  logging.Logger
+	activeBackgroundWorkers sync.WaitGroup
+}
+
+// PIDConfig is values needed to configure a PID control loop.
+type PIDConfig struct {
+	Type string  `json:"type,omitempty"`
+	P    float64 `json:"p"`
+	I    float64 `json:"i"`
+	D    float64 `json:"d"`
+}
+
+// Options contains values used for a control loop.
+type Options struct {
+	// PositionControlUsingTrapz
+	PositionControlUsingTrapz bool
+
+	// SensorFeedbackVelocityControl
+	SensorFeedbackVelocityControl bool
+
+	// DerivativeType
+	DerivativeType string
+
+	// UseCustomeConfig
+	UseCustomConfig bool
+
+	// CompleteCustomConfig
+	CompleteCustomConfig Config
+
+	// NeedsAutoTuning
+	NeedsAutoTuning bool
+
+	// LoopFrequency
+	LoopFrequency float64
+
+	// ControllableType
+	ControllableType string
+
+	// NeedsSingleAutoTuning
+	NeedsSingleAutoTuning bool
+}
+
+// SetupPIDControlConfig creates a control config.
+func SetupPIDControlConfig(
+	pidVals []PIDConfig,
+	componentName string,
+	options Options,
+	c Controllable,
+	logger logging.Logger,
+) (*PIDLoop, error) {
+	pidLoop := &PIDLoop{
+		Controllable: c,
+		logger:       logger,
+		Options:      options,
+		ControlConf:  Config{},
+		ControlLoop:  nil,
+	}
+
+	// set controlConf as either an optional custom config, or as the default control config
+	if options.UseCustomConfig {
+		pidLoop.ControlConf = options.CompleteCustomConfig
+		for i, b := range options.CompleteCustomConfig.Blocks {
+			if b.Type == blockSum {
+				sumIndex = i
+			}
+		}
+	} else {
+		pidLoop.createControlLoopConfig(pidVals, componentName)
+	}
+
+	// auto tune the control loop if needed
+	if options.NeedsAutoTuning {
+		cancelCtx, cancelFunc := context.WithCancel(context.Background())
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if options.NeedsSingleAutoTuning {
+		if err := pidLoop.StartControlLoop(); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
can this be moved into Options.NeedsAutoTuning? we already have a `p.Options.UseCustomConfig` case in there.

Reply:
I do think there's probably a better way to do this, but I was running into issues with the case where only one loop needed to be auto tuned. line linear is set but angular needs to be tuned

Reply:
very open to reworking this tho because I don't like it

Reply:
added a possible suggestion, but feel free to resolve this